paddle/fluid/framework/parallel_executor.cc

```diff
@@ -159,11 +170,21 @@ void ParallelExecutor::BCastParamsToGPUs(
           t->Resize(dims);
           buffer = t->mutable_data(place, main_tensor.type());
         }
-        auto &nccl_ctx = member_->nccl_ctxs_->at(place);
-        platform::dynload::ncclBcast(buffer, numel, data_type, 0,
-                                     nccl_ctx.comm_, nccl_ctx.stream());
+        buffers.push_back(buffer);
+      }
+
+      PADDLE_ENFORCE_EQ(member_->places_.size(), buffers.size(),
+                        "variables' buffer size to bcast NOT equal to places");
+      {
+        platform::NCCLGroupGuard guard;
+        for (size_t i = 0; i < member_->places_.size(); ++i) {
+          auto &nccl_ctx = member_->nccl_ctxs_->at(member_->places_[i]);
+          platform::dynload::ncclBcast(buffers[i], numel, data_type, 0,
+                                       nccl_ctx.comm_, nccl_ctx.stream());
+        }
+        member_->nccl_ctxs_->WaitAll();
```

Comment:
This line may not needed? since `ncclgroupend` will sync all group calls.

Reply:
@chengduoZH Can you please help take a look at this?

Reply:
@typhoonzero line 185 is necessary,
- `ncclgroupend` only ensures that there is only one thread can invoke nccl function, but not sync the calls.
- the invoking of nccl functions is async, so we need use `WaitAll()` to ensure the invoking have completed on GPU side.